src/Widgets/CameraView.vala

```diff
@@ -190,6 +246,62 @@ public class Camera.Widgets.CameraView : Gtk.Box {
         current_device = camera;
     }
 
+    private void create_pipeline_fd (int camera_fd) {
```

Comment:
Hi there. I've been looking into starting a new application involving webcam recording with GStreamer so I was taking a look through this source code and found this PR.

This function is largely duplicate code with the `create_pipeline` function which uses the `v4l2src` GStreamer element. The common code could be factored out to another function. However, I'm not sure it's really necessary to keep the V4L2 code path at all. From https://blogs.gnome.org/uraeus/2021/10/01/pipewire-and-fixing-the-linux-video-capture-stack/ :

> we do not recommend your GStreamer application using the v4l2 or libcamera plugins, instead we recommend that you use the PipeWire plugins

In the future you may be able to remove the explicit call to the XDG Portal too:

> There are a few major items we are still trying to decide upon in terms of the interaction between PipeWire and the [Camera portal API](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.Camera). It would be tempting to see if we can hide the Camera portal API behind the PipeWire API, or failing that at least hide it for people using the GStreamer plugin. That way all applications get the portal support for free when porting to GStreamer instead of requiring using the Camera portal API as a second step. On the other side you need to set up the screen sharing portal yourself, so it would probably make things more consistent if we left it to application developers to do for camera access too.

However, taking a look at the [source code for the pipewiresrc GStreamer element](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/gst/gstpipewiresrc.c), I don't see any usage of the XDG Portal. I'm unclear if that's still on the roadmap to be implemented or if that idea was dismissed. I asked in the Pipewire Matrix room earlier today but haven't gotten a response yet.

Reply:
[Cheese](https://gitlab.gnome.org/GNOME/cheese/-/blob/3b5eadacbe4bf21f380bf9d53ecae46224f83de0/libcheese/cheese-camera.c#L226) and [Kamoso](https://invent.kde.org/multimedia/kamoso/-/blob/9a98cf280734d8fb4e9ad6a4ad25f5c8d3f2798b/src/video/webcamcontrol.cpp#L281) use the `camerabin` GStreamer element. 

Pipewire [registers `pipewiresrc` with `GST_RANK_PRIMARY +1`](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/b99c71262ef64f138f3bb26b11b4e2ea94faa445/src/gst/gstpipewire.c#L49) so `camerabin` automatically uses `pipewiresrc` if it is available. Running `gst-launch-1.0 -v camerabin` confirms this:

```
/GstCameraBin:camerabin0/GstWrapperCameraBinSrc:camerasrc/GstAutoVideoSrc:camerasrc-real-src/GstPipeWireSrc:camerasrc-real-src-actual-src-pipewir.GstPad:src: caps = video/x-raw, format=(string)YUY2, width=(int)1280, height=(int)720, framerate=(fraction)10/1
/GstCameraBin:camerabin0/GstWrapperCameraBinSrc:camerasrc/GstAutoVideoSrc:camerasrc-real-src.GstGhostPad:src: caps = video/x-raw, format=(string)YUY2, width=(int)1280, height=(int)720, framerate=(fraction)10/1
```

Without Pipewire's GStreamer plugin, `camerabin` uses V4L2, running `GST_PLUGIN_FEATURE_RANK=pipewiresrc:NONE gst-launch-1.0 -v camerabin`:

```
/GstCameraBin:camerabin0/GstWrapperCameraBinSrc:camerasrc/GstAutoVideoSrc:camerasrc-real-src/GstV4l2Src:camerasrc-real-src-actual-src-v4l.GstPad:src: caps = video/x-raw, width=(int)1280, height=(int)720, framerate=(fraction)10/1, format=(string)YUY2, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive, colorimetry=(string)2:4:5:1
```

So `camerabin` would support both the cases of Pipewire being present or not. However, I don't know if there's a way to pass the file descriptor from the portal to `camerabin`.